### DISCUSSION TITLE
How to change render resolution?

### DISCUSSION THREAD

**Discussion by paulorrfilho** [google/filament#4790] [Q&A]
How do I change the rendering resolution to improve performance?
I have seen this being suggested in previous discussions and I tried adding the following lines to my code. Rendering quality is clearly affected but I don't see much improvement in terms of animation smoothness.

Am I doing this the right way?

```
surfaceView?.holder?.surfaceFrame?.let {
     uiHelper.setDesiredSize((it.width() * 0.60).toInt(), (it.height() * 0.60).toInt())
}`

**Comment by paulorrfilho** [comment]
@romainguy , would you mind commenting on this?
This has been suggested in previous discussions but I don't see any performance improvements, even with dynamic resolution turned on. Maybe I am doing this the wrong way?

**Reply by romainguy** [reply]
That's the proper way to do it. Dynamic resolution would also help. Your performance issue might come from something else then (CPU? Poor frame pacing? etc.).

**Reply by paulorrfilho** [reply]
CPU usage is surprisingly low (25%-35%), even for low spec phones, however, FPS seems to be around 15-25 and it is stuttering.
RAM usage is high (about 800MB) due to "graphics", could this be the problem?

My app has 5 ModelViewers (and 5 engines) running at the same time, one in each fragment. Not sure if this can lead to problems. 

Medium/High-end devices run smoothly, no problem. Low-end devices, specially PowerVR GE8320 (the most popular GPU in Asian/Latin American countries), have the issues I mentioned.

What puzzles me is that it does not matter the resolution I choose, the lag is the same.

**Reply by romainguy** [reply]
Using so many engines will increase RAM usage. You can use only one engine and have multiple renderers share it. That would help. If changing resolution doesn't help then the issue will be something else like number of vertices. Older IMG GPUs can greatly suffer from going above a certain number of vertices.

**Reply by paulorrfilho** [reply]
Ok, I was able to create only one engine and RAM usage is much smaller, however, lag is much worse. Each fragment has its own render now linked to a single engine.
I think the issue is related to the multiple swap chains, uiHelpers, and displayHelpers, should each render have an exclusive swap chain or should I also have just one?

**Reply by romainguy** [reply]
Are all those renderers rendering at the same time? You might be witness the cost of swapping render targets in the GL context. It very much sounds like a CPU bottleneck. How do you check CPU usage?

BTW here's a demo app I wrote that uses a single Engine but multiple Renderers/SwapChains: https://github.com/romainguy/sample-materials-shop/ (and they all render at the same time). It also shares various objects across renderers (like the indirect light).

**Reply by paulorrfilho** [reply]
I am checking the Android Studio Profiler.
ok, I will take a look at that project.
thank you

**Reply by paulorrfilho** [reply]
and yes, they are all rendering at the same time

**Reply by paulorrfilho** [reply]
So, I structure my code to look similar to the demo you mentioned (shared engine, material, indirect light, asset loader, etc). Performance is much worse and proportional to the number of fragments being rendered. It is as if using multiple engines provides some sort of threading while having just one engine the work is done serially. Does this make sense?
